[DOCS] Fix reranking IA, move retrievers to search api overview (#112949)
elastic/elasticsearch: docs/reference/reranking/index.asciidoc

[[reranking-overview]]
= Reranking

> Rerankers improve the relevance of results from earlier-stage retrieval mechanisms.

Reviewer: I feel like we could add a sentence or two here to make it a bit more approachable to people who don't know what `earlier-stage retrieval mechanisms` or `first-stage rerankers` are. At the bare minimum, perhaps talk about generating a set of candidates that then get funneled into a late stage reranker that performs more computationally expensive reranking tasks? 

Perhaps this could be an opportunity for a diagram or a link to a more in-depth post in Search Labs (if one exists)?

Author: > I feel like we could add a sentence or two here to make it a bit more approachable to people who don't know what earlier-stage retrieval mechanisms or first-stage rerankers are.

💯

First-stage rankers must be very fast and efficient because they process either the entire corpus, or all matching documents.

In a multi-stage pipeline, you can progressively use more computationally intensive ranking functions and techniques, as they will operate on smaller result sets at each step.
This helps avoid query latency degradation and keeps costs manageable.

> Currently, the following reranking techniques are available in {es}:

Reviewer: Could be worth linking out to RRF as an alternate solution here?

[[re-ranking-overview]]
= Re-ranking

Many search systems are built on two-stage retrieval pipelines.

The first stage uses cheap, fast algorithms to find a broad set of possible matches.

The second stage uses a more powerful model, often machine learning-based, to reorder the documents.
This second step is called re-ranking.
Because the resource-intensive model is only applied to the smaller set of pre-filtered results, this approach returns more relevant results while still optimizing for search performance and computational costs.

{es} supports various ranking and re-ranking techniques to optimize search relevance and performance.

[float]
[[re-ranking-two-stage-pipeline]]
== Two-stage retrieval pipelines

[float]
[[re-ranking-first-stage-pipeline]]
=== First stage: initial retrieval

[float]
[[re-ranking-ranking-overview-bm25]]
==== Full-text search: BM25 scoring

{es} ranks documents based on term frequency and inverse document frequency, adjusted for document length.

> https://en.wikipedia.org/wiki/Okapi_BM25[BM25] is the default statistical scoring algorithm in {es} and works out-of-the-box.

Reviewer: ```suggestion
https://en.wikipedia.org/wiki/Okapi_BM25[BM25] is the default statistical scoring algorithm in {es}.
```

[float]
[[re-ranking-ranking-overview-vector]]
==== Vector search: similarity scoring

Vector search involves transforming data into dense or sparse vector embeddings to capture semantic meanings, and computing similarity scores for query vectors.
Store vectors using `semantic_text` fields for automatic inference and vectorization or `dense_vector` and `sparse_vector` fields when you need more control over the underlying embedding model.
Query vector fields with `semantic`, `knn` or `sparse_vector` queries to compute similarity scores.
Refer to <<semantic-search,semantic search>> for more information.

[float]
[[re-ranking-ranking-overview-hybrid]]
==== Hybrid techniques

Hybrid search techniques combine results from full-text and vector search pipelines.
{es} enables combining lexical matching (BM25) and vector search scores using the <<rrf,Reciprocal Rank Fusion (RRF)>> algorithm.

[float]
[[re-ranking-overview-second-stage]]

> === Second Stage: Re-ranking

Author: ```suggestion
=== Second stage: Re-ranking
```

[[re-ranking-overview]]

> = Re-ranking

Reviewer: I'm noticing some inconsistency - sometimes we use `Reranking` and other times we use `Re-ranking`. I slightly prefer `reranking` but am fine with either, though we should do a pass for consistency.

Author: realized that in course of this PR and tried to align on hyphenated version, will need to update the semantic re-reranking page too

Author: FWIW I also prefer `reranking` but the hyphenated version seems more au courant :)

Many search systems are built on two-stage retrieval pipelines.

> The first stage uses cheap, fast algorithms to find a subset of possible matches.

Reviewer: ```suggestion
The first stage uses cheap, fast algorithms to find a broad set of possible matches.
```

The second stage uses a more powerful model, often machine learning-based, to reorder the documents.
This second step is called re-ranking.

> This approach balances computational costs, because the resource-intensive model is only applied to a smaller set of pre-filtered results.

Reviewer: ```suggestion
Because the resource-intensive model is only applied to the smaller set of pre-filtered results, this approach returns more relevant results while still optimizing for search performance and computational costs.
```

> {es} supports various ranking and re-ranking techniques to optimize search relevance and performance.

Reviewer: Question: Should this end with a `.` or `:`?

Author: `.`

[float]
[[re-ranking-two-stage-pipeline]]
== Two-stage retrieval pipelines

> Learn about retrieval pipelines and how re-ranking can be applied to your existing search experience.

Reviewer: Starting this sentence with `Learn about` this deep into the document feels a bit awkward. Do we need it?

Author: Yeah definitely a placeholder kinda glue sentence...I'll ponder alternative

Author: let's remove it entirely

> Learn about retrieval pipelines and how re-ranking can be applied to your existing search experience.

Author: ```suggestion
```

[float]
[[re-ranking-first-stage-pipeline]]
=== First stage: initial retrieval

[float]
[[re-ranking-ranking-overview-bm25]]

> ==== Full-text search: BM25 scoring

Reviewer: I've commonly seen questions on discuss surrounding BM25. It's old hat to search veterans, but new learners could benefit from a link here to learn more about BM25.

Author: done

{es} ranks documents based on term frequency and inverse document frequency, adjusted for document length.

> BM25 is the default statistical scoring algorithm in {es} and works out-of-the-box.

Author: ```suggestion
https://en.wikipedia.org/wiki/Okapi_BM25[BM25] is the default statistical scoring algorithm in {es} and works out-of-the-box.
```

[float]
[[re-ranking-ranking-overview-vector]]
==== Vector search: similarity scoring

Vector search involves transforming data into dense or sparse vector embeddings to capture semantic meanings, and computing similarity scores for query vectors.

> Store vectors using `semantic` fields for automatic vectorization or `dense_vector` and `sparse_vector` fields when you need more control over the underlying embedding model.

Reviewer: ```suggestion
Store vectors using `semantic_text` fields for automatic inference and vectorization or `dense_vector` and `sparse_vector` fields when you need more control over the underlying embedding model.
```

We should link to each mapping page here too.

Author: but I will take the corrections ;-)

> Query vectors with `semantic` queries or `knn` and `sparse_vector` queries to compute similarity scores.

Reviewer: ```suggestion
Query vector fields with `semantic`, `knn` or `sparse_vector` queries to compute similarity scores.
```

Links to the query DSL pages for each query type would be helpful here. Also, I think we want to start specifying "retriever first", so maybe we should mention that here before we mention the queries?

Refer to <<semantic-search,semantic search>> for more information.

[float]
[[re-ranking-ranking-overview-hybrid]]
==== Hybrid techniques

> Hybrid search techniques combine results from full-text and vector search pipelines.

Reviewer: I am not sure how much we have here, but we've been getting more and more questions and feedback on people who want more information on linear boosting as well. I wonder if a sentence here and a link to somewhere else (if we have it) would be worth doing.

Author: In future we want to have fuller docs sections for:

- Full-text search
- ML-powered search
- Hybrid search

I think we'll want to revisit that when the time comes. The goal here is to put LTR and semantic reranking in a smarter IA, first and foremost, not to be the exhaustive source of truth for all these topics.

Reviewer: Sure sure, push back on my scope creep 😀

> {es} enables combining lexical matching (BM25) and vector search scores using the advanced <<rrf,Reciprocal Rank Fusion (RRF)>> algorithm.

Reviewer: ```suggestion
{es} enables combining lexical matching (BM25) and vector search scores using the <<rrf,Reciprocal Rank Fusion (RRF)>> algorithm.
```

[float]
[[re-ranking-overview-second-stage]]
=== Second Stage: Re-ranking

When using the following advanced re-ranking pipelines, first-stage retrieval mechanisms effectively generate a set of candidates.
These candidates are funneled into the re-ranker to perform more computationally expensive re-ranking tasks.

[float]
[[re-ranking-overview-semantic]]
==== Semantic re-ranking

<<semantic-reranking>> uses machine learning models to reorder search results based on their semantic similarity to a query.
Models can be hosted directly in your {es} cluster, or you can use <<inference-apis,inference endpoints>> to call models provided by third-party services.

> Enables out-of-the-box semantic search capabilities on existing full-text search indices.

Reviewer: ```suggestion
Semantic re-ranking out-of-the-box semantic search capabilities on existing full-text search indices.
```

[float]
[[re-ranking-overview-ltr]]
==== Learning to Rank (LTR)

<<learning-to-rank>> is for advanced users.

> Train a machine learning model to build a ranking function for your search experience that updates over time.
> Best suited for when you have ample training data and need highly customized relevance tuning.

Reviewer: ```suggestion
Learn to rank involves training a machine learning model to build a ranking function for your search experience that updates over time.
LTR is best suited for when you have ample training data and need highly customized relevance tuning.
```
